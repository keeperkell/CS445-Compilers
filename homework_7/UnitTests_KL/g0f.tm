* C- compiler version C-S21
* Built: Apr 18, 2021 (toffset telemetry)
* Author: Robert B. Heckendorn
* File compiled:  g0f.c-
* 
* ** ** ** ** ** ** ** ** ** ** ** **
* FUNCTION input
  1:     ST  3,-1(1)	Store return address 
  2:     IN  2,2,2	Grab int input 
  3:     LD  3,-1(1)	Load return address 
  4:     LD  1,0(1)	Adjust fp 
  5:    JMP  7,0(3)	Return 
* END FUNCTION input
* 
* ** ** ** ** ** ** ** ** ** ** ** **
* FUNCTION output
  6:     ST  3,-1(1)	Store return address 
  7:     LD  3,-2(1)	Load parameter 
  8:    OUT  3,3,3	Output integer 
  9:     LD  3,-1(1)	Load return address 
 10:     LD  1,0(1)	Adjust fp 
 11:    JMP  7,0(3)	Return 
* END FUNCTION output
* 
* ** ** ** ** ** ** ** ** ** ** ** **
* FUNCTION inputb
 12:     ST  3,-1(1)	Store return address 
 13:    INB  2,2,2	Grab bool input 
 14:     LD  3,-1(1)	Load return address 
 15:     LD  1,0(1)	Adjust fp 
 16:    JMP  7,0(3)	Return 
* END FUNCTION inputb
* 
* ** ** ** ** ** ** ** ** ** ** ** **
* FUNCTION outputb
 17:     ST  3,-1(1)	Store return address 
 18:     LD  3,-2(1)	Load parameter 
 19:   OUTB  3,3,3	Output bool 
 20:     LD  3,-1(1)	Load return address 
 21:     LD  1,0(1)	Adjust fp 
 22:    JMP  7,0(3)	Return 
* END FUNCTION outputb
* 
* ** ** ** ** ** ** ** ** ** ** ** **
* FUNCTION inputc
 23:     ST  3,-1(1)	Store return address 
 24:    INC  2,2,2	Grab char input 
 25:     LD  3,-1(1)	Load return address 
 26:     LD  1,0(1)	Adjust fp 
 27:    JMP  7,0(3)	Return 
* END FUNCTION inputc
* 
* ** ** ** ** ** ** ** ** ** ** ** **
* FUNCTION outputc
 28:     ST  3,-1(1)	Store return address 
 29:     LD  3,-2(1)	Load parameter 
 30:   OUTC  3,3,3	Output char 
 31:     LD  3,-1(1)	Load return address 
 32:     LD  1,0(1)	Adjust fp 
 33:    JMP  7,0(3)	Return 
* END FUNCTION outputc
* 
* ** ** ** ** ** ** ** ** ** ** ** **
* FUNCTION outnl
 34:     ST  3,-1(1)	Store return address 
 35:  OUTNL  3,3,3	Output a newline 
 36:     LD  3,-1(1)	Load return address 
 37:     LD  1,0(1)	Adjust fp 
 38:    JMP  7,0(3)	Return 
* END FUNCTION outnl
* 
* ** ** ** ** ** ** ** ** ** ** ** **
* FUNCTION cat
* TOFF set: -3
 39:     ST  3,-1(1)	Store return address 
* COMPOUND
* TOFF set: -3
* Compound Body
* IF
 40:     LD  3,-2(1)	Load variable x
 41:     ST  3,-3(1)	Push left side 
* TOFF dec: -4
 42:    LDC  3,0(6)	Load integer constant 
* TOFF inc: -3
 43:     LD  4,-3(1)	Pop left into ac1 
 44:    TGT  3,4,3	Op > 
* THEN
* RETURN
 46:     LD  3,-2(1)	Load variable x
 47:     ST  3,-3(1)	Push left side 
* TOFF dec: -4
 48:     LD  3,-2(1)	Load variable x
* TOFF inc: -3
 49:     LD  4,-3(1)	Pop left into ac1 
 50:    ADD  3,4,3	Op + 
 51:    LDA  2,0(3)	Copy result to return register 
 52:     LD  3,-1(1)	Load return address 
 53:     LD  1,0(1)	Adjust fp 
 54:    JMP  7,0(3)	Return 
 45:    JZR  3,10(7)	Jump around the THEN if false [backpatch] 
* ELSE
* EXPRESSION
 56:     LD  3,-2(1)	Load variable x
 57:     ST  3,-3(1)	Push left side 
* TOFF dec: -4
 58:     LD  3,-2(1)	Load variable x
* TOFF inc: -3
 59:     LD  4,-3(1)	Pop left into ac1 
 60:    MUL  3,4,3	Op * 
 55:    JMP  7,5(7)	Jump around the ELSE [backpatch] 
* END IF
* TOFF set: -3
* END COMPOUND
* Add standard closing in case there is no return statement
 61:    LDC  2,0(6)	Set return value to 0 
 62:     LD  3,-1(1)	Load return address 
 63:     LD  1,0(1)	Adjust fp 
 64:    JMP  7,0(3)	Return 
* END FUNCTION cat
* 
* ** ** ** ** ** ** ** ** ** ** ** **
* FUNCTION main
* TOFF set: -2
 65:     ST  3,-1(1)	Store return address 
* COMPOUND
* TOFF set: -2
* Compound Body
* EXPRESSION
* CALL output
 66:     ST  1,-2(1)	Store fp in ghost frame for output
* TOFF dec: -3
* TOFF dec: -4
* Param 1
* CALL cat
 67:     ST  1,-4(1)	Store fp in ghost frame for cat
* TOFF dec: -5
* TOFF dec: -6
* Param 1
 68:    LDC  3,88(6)	Load integer constant 
 69:     ST  3,-6(1)	Push parameter 
* TOFF dec: -7
* Param end cat
 70:    LDA  1,-4(1)	Ghost frame becomes new active frame 
 71:    LDA  3,1(7)	Return address in ac 
 72:    JMP  7,-34(7)	CALL cat
 73:    LDA  3,0(2)	Save the result in ac 
* Call end cat
* TOFF set: -4
 74:     ST  3,-4(1)	Push left side 
* TOFF dec: -5
* CALL cat
 75:     ST  1,-5(1)	Store fp in ghost frame for cat
* TOFF dec: -6
* TOFF dec: -7
* Param 1
 76:    LDC  3,17(6)	Load integer constant 
 77:    NEG  3,3,3	Op unary - 
 78:     ST  3,-7(1)	Push parameter 
* TOFF dec: -8
* Param end cat
 79:    LDA  1,-5(1)	Ghost frame becomes new active frame 
 80:    LDA  3,1(7)	Return address in ac 
 81:    JMP  7,-43(7)	CALL cat
 82:    LDA  3,0(2)	Save the result in ac 
* Call end cat
* TOFF set: -5
* TOFF inc: -4
 83:     LD  4,-4(1)	Pop left into ac1 
 84:    ADD  3,4,3	Op + 
 85:     ST  3,-4(1)	Push parameter 
* TOFF dec: -5
* Param end output
 86:    LDA  1,-2(1)	Ghost frame becomes new active frame 
 87:    LDA  3,1(7)	Return address in ac 
 88:    JMP  7,-83(7)	CALL output
 89:    LDA  3,0(2)	Save the result in ac 
* Call end output
* TOFF set: -2
* TOFF set: -2
* END COMPOUND
* Add standard closing in case there is no return statement
 90:    LDC  2,0(6)	Set return value to 0 
 91:     LD  3,-1(1)	Load return address 
 92:     LD  1,0(1)	Adjust fp 
 93:    JMP  7,0(3)	Return 
* END FUNCTION main
  0:    JMP  7,93(7)	Jump to init [backpatch] 
* INIT
 94:    LDA  1,0(0)	set first frame at end of globals 
 95:     ST  1,0(1)	store old fp (point to self) 
* INIT GLOBALS AND STATICS
* END INIT GLOBALS AND STATICS
 96:    LDA  3,1(7)	Return address in ac 
 97:    JMP  7,-33(7)	Jump to main 
 98:   HALT  0,0,0	DONE! 
* END INIT
