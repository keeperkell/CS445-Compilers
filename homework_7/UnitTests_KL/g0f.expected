Number of warnings: 0
Number of errors: 0
------------------------------------
Loading file: UnitTests/g0f.tm
176
Bye.
